### ISSUE TITLE
Discussion: Plugins for admin GUI widgets

### ISSUE THREAD

**Issue by davidferguson** [lektor/lektor#665]
This is similar to #217.

One obstacle I've found with Lektor so far is not being able to add custom input widgets to the admin GUI. The site I'm currently developing requires a few custom widgets and while we're currently just manually adding them in, it would be great to give plugins this ability.

What I'm currently thinking is that a widget plugin would be a React class, that gets inserted into `widgets.jsx`, and is therefore available for use by creating a custom type plugin and setting the `widget` option to the name of the class.

What I'm currently wondering is: The current plugin system is (obviously) in Python, and doesn't do anything for the js admin UI. Would it be better to try and "bridge-the-gap" between them (so to speak) and have plugin events for the admin UI (and then have to worry about how a python plugin is going to interact with js events), or, have a separate plugin thing for the admin gui?

I'm leaning towards having a separate admin GUI plugin system, as this would be much much easier to implement, and makes more sense to me. I'm thinking something like [architect](https://www.npmjs.com/package/architect) would be applicable here, and allow a `widgets` directory to be populated with user-created widgets, which get loaded into the admin GUI. I'll need to see how this interacts with webpack though.

This would help with #663, #661 and #7.

I'd appreciate thoughts on this before I start doing any work @runfalk @nixjdm

**Comment by runfalk** [comment]
Yeah, I think GUI widget plugins is something we eventually want to allow. The tricky part is that no-one is happy about the current admin frontend codebase (#457, #458), and as soon as we allow GUI plugins we get a bit locked in to the current codebase.

I'm not sure how you intend the events to work, do you want plugins to define hooks of some kind? Could you elaborate a bit with an example here?

I think it's best to make GUI plugins part of normal plugins so they can be installed the same way, and that we don't end up with situations where you need to install both a Python and GUI plugin separately. It is possible to bundle data files (js) as part of Python wheels so I don't think it will be an issue for distribution. Something like architect can still work with this solution.

The gist of it is that I'm happy to allow GUI plugins, as long as we don't build ourselves into a corner that prevents us from throwing away the current admin.

**Comment by nixjdm** [comment]
@runfalk I feel like I could have typed that myself. I echo everything you said.

FWIW about the packaging, if we nail down how frontend plugins would work, we can try make some boilerplate plugin code as visible as possible to make creating this kind of plugin easier. If it makes sense we could also include some supporting code in core to simplify it.

**Comment by uwekamper** [comment]
I would like to suggest something here.

Right now the different "modules" of the Lektor (e.g. api.py, see: https://github.com/lektor/lektor/tree/master/lektor/admin/modules) 
are registered at this point: https://github.com/lektor/lektor/blob/master/lektor/admin/modules/__init__.py#L4 . The `register_modules` method basically adds the URLs of each module to the Flask App-object.

A plugin cannot add its own "module" to this list.

If a plugin could register its own "module" through some kind of hook, any plugin could effectively get its own URL namespace below http://localhost:5000/admin/. For example http://localhost:5000/admin/plugins/my_crazy_plugin/

A hook that would allow a plugin to register its own URLs would look like this:

```
import my_crazy_plugin.CrazyAdmin

class MyCrazyPlugin(Plugin):
    name = 'My Crazy Plugin'
    description = 'This is a crazy plugin for testing purposes.'

    def on_setup_urls(self, admin_url_modules):
        admin_url_modules.append('my_crazy_plugin', my_crazy_plugin.CrazyAdmin)
```

I'd be willing to contribute a pull-request if you think this is worth it.

**Comment by runfalk** [comment]
The main problem I see here is how the plugin visually integrate with the rest of Lektor. Should these plugins to be completely standalone? Should they use the same GUI as the rest of the admin? The admin GUI was recently changed from JS to TypeScript. This kind of change would not be possible unless these modules are completely stand alone. Alternatively we allow Lektor to break plugins each release, but end users would probably not be happy about that.

I'm not sure exactly what you plan on putting in these modules, maybe if you could provide a use case here?

**Comment by uwekamper** [comment]
@runfalk Thanks for reacting :-)

> The main problem I see here is how the plugin visually integrate with the rest of Lektor.

In a first step there would be no GUI integration at all. Lektor would already be twice as useful to me – even without direct integration into the admin.

For the future it would be nice to be able to integrate such plugin into
I do not have a design degree but I tried to draw some mock-ups, please see below.

> Should these plugins to be completely standalone? Should they use the same GUI as the rest of the admin? 

These plugins would generate their own HTML and would only use the Lektor server to deliver the HTML.

## Use case: Oauth2 workflow and access to external APIs

I have a a customer who has a Facebook page and regularly posts updates about the business on Facebook.
I would like to create a plugin that displays these posts as business's front page. In order to do that my Lektor plugin needs an  authentication token. To get an authentication token from Facebook my customer needs to complete an OAuth workflow in **their** browser (because only the page owner can access the necessary APIs), the workflow would be as follows:

- Customer opens Lektor Admin
- Customer directs browser to localhost:5000/admin/plugins/my_facebook/
- The plugin presents a "Login with Facebook" button
- Customer clicks on the "Login with Facebook" button and is redirected through the Facebook OAuth2 flow and allows the plugin to access the Facebook API.
- Facebook redirects browser back to  localhost:5000/admin/plugins/my_facebook/callback?token=xxxxxxxxxxxxxxxxxxx
- The plugins stores the token locally and can now access the Facebook API to insert a list of Facebook posts into /index.html 

The same basic workflow could be used with other services that use OAuth2: 

- Twitter, Instagram and other social networks
- Access a Google spreadsheets and display them on a Lektor static page as HTML
- Access a service like Airtable.
- Access to an internal API of a business
- ...

**Comment by uwekamper** [comment]
Here are some mockups / ideas on how these Plugins could be integrated in the Lektor admin UI

## Variant 1: New sub-menu on the left side

Here is what the plugin code might look like (please consider this as pseudocode):

```python
class FacebookPostsPlugin(Plugin):
     def on_setup_urls(self, admin_url_modules):
         admin_url_modules.append('facebook-posts-plugin', my_crazy_plugin.CrazyAdmin)
     def on_server_startup(env, **extras):
        env.register_plugin_button("Facebook", "/admin/plugins/facebook-posts-plugin")
```

The Lektor GUI would then draw a button that looks like this:
 
<img width="1364" alt="plugin-button-1" src="https://user-images.githubusercontent.com/829305/106652729-2d090e00-6596-11eb-8d40-d7b18bd8f5bd.png">

## Variant 2: Display registered links inside the Page actions menu

Pseudo code example:

```python
class FacebookPostsPlugin(Plugin):
     [...]

     def on_server_startup(env, **extras):
        env.register_page_action("Connect to Facebook", "/admin/plugins/facebook-posts-plugin")
```

Result:

![image](https://user-images.githubusercontent.com/829305/106653815-95a4ba80-6597-11eb-8954-33343e30a79d.png)

## Variant 3: Add a plugin menu button on the upper right side

Again, the Plugin needs to register itself with Lektor

The result is an extra menu button on the top right of the Lektor admin (the jigsaw puzzle piece icon). Clicking on the jigsaw piece opens up a modal dialog that displays the list of Plugins and plugin links:

<img width="1364" alt="plugin-button-2" src="https://user-images.githubusercontent.com/829305/106655343-83c41700-6599-11eb-8d0a-5a0c15ae9bfd.png">

**Comment by davidferguson** [comment]
I was mid-way typing out a reply to this, but @uwekamper has already discussed much of what I was going to say!

We use Lektor for a large website, and are happy with it, however one area that is not so good is the GUI. Our goal is to have the system used almost entirely by non-technical users, including:
- editing (which is currently possible with the existing GUI)
- model editing
- template editing
- static assets upload
- config file editing
- plugin installation
- a review system, so edits can be reviewed before being published (this is obviously quite challenging, as it requires some form of authentication/authorisation which is a big change to the existing GUI)

The current GUI also isn't great for large sites. For the most part it works, but there are parts that don't, such as the sidebar for subpages which has no search or attachments which has no pagination (and we have pages with thousands of attachments!), or the publish dialog which is synchronous and therefore doesn't work when the build takes over an hour, as it does with our site!

To this end, I did think about creating a entirely new GUI for Lektor, earlier this month. I really only spent a couple of days on it, and it's far from even being feature complete compared to the current GUI, let alone all the new things I want to add to it! I'm also not a UI designer, nor familiar with React Router, so it could be mostly garbage, but it's based on feedback from the couple of people who currently edit our site along with my observations. I've included a couple of screenshots of what it currently looks like (again it's very rough!) but if there's any interest at all I'll put it on GitHub for people to check out.

![01-home-screen](https://user-images.githubusercontent.com/13520342/106658668-7b69dd00-6595-11eb-8021-ebc8f5d13b68.png)
----
![02-page-screen](https://user-images.githubusercontent.com/13520342/106658675-7c9b0a00-6595-11eb-995a-ca867b644d8e.png)
----
![03-edit-screen](https://user-images.githubusercontent.com/13520342/106658679-7dcc3700-6595-11eb-896a-adda333116b0.png)


----

Going back to plugins, how I'd envisioned they'd work in my GUI (if I ever finished it!) was similar to what @uwekamper has outlined above. There would be several types of plugin, including:
1. Full screen plugin, that has a menu option in the left menu
2. Editing Widget plugin, of a new React Component that is a editor widget
3. Maybe various hooks in other parts of the GUI, such as a widget for the home screen, or extra fields for a publisher that has user configurable paramaters

Sorry about this big wall of text that isn't all related to plugins, but I just thought I'd share our use use of Lektor, and my thoughts on how I want plugins to work.

**Comment by uwekamper** [comment]
@davidferguson Just a side note: It is *big* feature for me that the current admin can only be used to edit content and upload files. If the new admin also allows model editing and config file editing, this would diminish the Lektor admin's usefulness a lot – because I would not be able to give this to end users. Because then my end users would be able to change a config und thus break the whole site.

**Comment by runfalk** [comment]
@uwekamper I can kind of see what you're going for. I'd say that such features are unlikely to be included in mainline Lektor at this point. The list of changes we might want to do with the admin:

* Redesign it for a cleaner look
* Fix performance issues with larger sites
* Better feedback when builds fail
* Allow editing things like databags, templates and plugins (and maybe even models)
* Simplify hosting of the admin UI, which may include access control

We don't want to add admin GUI plugin support when we can't guarantee the stability of the API. Publishing the plugin hooks you mentioned may interfere with out ability to do these without breaking user installations.

If you're just interested in supporting the OAuth workflow you mentioned you could do so today by making a new package that depends on Lektor and copy the `run_server` implementation (https://github.com/lektor/lektor/blob/master/lektor/devserver.py#L105) to add as many Flask modules as you'd like.

You can also override parts of the admin by replacing https://github.com/lektor/lektor/blob/master/lektor/admin/modules/dash.py in your new package. If you're really keen on having these features I think this is a much quicker path to getting them. If we change the GUI and add the plugin hooks you can always transfer to that when the time comes.

Since you wouldn't fork Lektor you could still benefit from the occasional new release without too much effort.

@davidferguson Your screenshots is very similar to what I envision the admin should eventually be like, at least content wise. It should be possible to edit most things or at least view them from there. It would be especially nice to have the option to make model changes directly in the UI, since this is not actually possible without editing every content file by hand.

If we ever add access control I guess we could put permissions into the mix to disallow some users from changing the templates or deeper configuration of the site.

**Comment by uwekamper** [comment]
@runfalk I am having some difficulty understanding what you are suggesting. 

Here is what I thing I must do to add another endpoint to the Lektor admin webserver:

- create a virtualenv
- pip install Lektor into the virtualenv
- create a file called my_run_server.py
- copy the run_server part that you linked to, adding the additional modules to the Flask App
- instead of `lektor runserver` I then must use `python my_run_server.py` to access the Lektor admin.

Am I right with this (or at least in the right direction)? Or did you mean something completely different?

**Comment by uwekamper** [comment]
@runfalk I made a quick-n-dirty experimental setup on how to implement your idea with `run_server`:

https://github.com/uwekamper/monkey-see-monkey-patch

In order to call run_server ( see here: https://github.com/uwekamper/monkey-see-monkey-patch/blob/master/modified_lektor.py#L124 ), I also had to copy the function `server_cmd` from the `lektor.cli` module.

When running `python ./modified_lektor.py server` you can then browse to http://localhost:5000/facebook/hello which will return a simple "Hello World" message.

**Comment by runfalk** [comment]
Yeah, I was thinking that you could make a proper Python project with a `setup.py` and everything (that way you can pin Lektor to a specific version and it's easier to install. But what you just did works just as well when prototyping.

Be mindful that this is still a hack and that it can break when Lektor is updated (I don't think we guarantee API stability for internals).

Do you think this is enough to cover your use case?

**Comment by uwekamper** [comment]
@runfalk This is definitely enough to cover my use case - I can build a Facebook/Instagram integration *now* instead of waiting for the Lektor to provide an API for this.

I still think that Lektor should (in the future) provide an API for Plugins to register Blueprints during the startup phase of the dev server. Maybe we should discuss this in a separate GitHub issue, because this is not related to the admin UI code.

**Comment by uyar** [comment]
>  I've included a couple of screenshots of what it currently looks like (again it's very rough!) but if there's any interest at all I'll put it on GitHub for people to check out.

@davidferguson It's been two years and I just saw this but I'd be interested if the offer still stands.

**Comment by uyar** [comment]
As a side note, I have started developing an alternative admin panel: [uyar/lektor-tekir](/uyar/lektor-tekir).
